test/distributed/test_c10d.py

```diff
@@ -2389,6 +2388,62 @@ def run_and_verify_grad(model):
         )
         run_and_verify_grad(gpu_model)
 
+    @requires_gloo()
+    @skip_if_lt_x_gpu(2)
+    def test_find_unused_parameters_when_unused_parameters_empty(self):
+        """
+        An empty unused_parameters array does not imply find_unused_parameters =
+        false. This test makes sure that DDP allreduces unused parameters
+        accordingly where the forward pass in some process uses all parameters.
+        This unit test creates a module that uses all parameters in rank = 0, and
+        has unused parameters in other ranks.
+        """
+        class FindUnusedParamModule(nn.Module):
+            def __init__(self):
+                super(FindUnusedParamModule, self).__init__()
+                self.t0 = Task()
+                self.t1 = Task()
+
+            def task_parameters(self):
+                return (self.t0.p, self.t1.p)
+
+            def forward(self, x, rank):
+                return self.t1(self.t0(x)) if rank == 0 else self.t1(x)
+
+        def run_and_verify_grad(model):
+            # Run forward
+            output = model(8, self.rank)
+
+            # The grads of all parameters should be None at this point.
+            [self.assertIsNone(t_p.grad) for t_p in model.module.task_parameters()]
+
+            # Run backward
+            output.mean().backward()
+
+            # Now locally unused parameter should have grad updated on all ranks.
+            [self.assertIsNotNone(t_p.grad) for t_p in model.module.task_parameters()]
```

Comment:
nit: I wonder if we should also verify that the grad values are as expected? We can do this by 

1. creating two local models, one's `forward` uses both `t0` and `t1`, and the other one's `forward` only uses `t1`
2. run forward/backward on them, and then compute the mean grads of these two models
3. Compare the mean grad with DDP's grad. 

I noticed the test above (`test_global_local_unused_params_grad`) didn't test the accuracy either. So feel free to do this in a followup PR.

Reply:
I noted this and do a followup soon.